src/util/codec/convert.rs

```diff
@@ -133,7 +130,74 @@ fn get_valid_float_prefix(s: &str) -> &str {
         }
     }
 
-    if valid_len == 0 { "0" } else { &s[..valid_len] }
+    try!(handle_truncate(ctx, valid_len < s.len()));
+    if valid_len == 0 {
+        Ok("0")
+    } else {
+        Ok(&s[..valid_len])
+    }
+}
+
+/// It converts a valid float string into valid integer string which can be
+/// parsed by `i64::from_str`, we can't parse float first then convert it to string
+/// because precision will be lost.
+fn float_str_to_int_string(valid_float: &str) -> Result<String> {
+    let mut dot_idx: i64 = -1;
+    let mut e_idx: i64 = -1;
+    for (i, c) in valid_float.chars().enumerate() {
+        match c {
+            '.' => dot_idx = i as i64,
+            'e' | 'E' => e_idx = i as i64,
+            _ => (),
+        }
+    }
+
+    if dot_idx == -1 && e_idx == -1 {
+        return Ok(valid_float.to_owned());
+    }
+
+    let mut int_cnt;
+    let mut digits = String::with_capacity(valid_float.len());
+
+    if dot_idx == -1 {
+        digits.push_str(&valid_float[..e_idx as usize]);
+        int_cnt = digits.len() as i64;
+    } else {
+        digits.push_str(&valid_float[..dot_idx as usize]);
+        int_cnt = digits.len() as i64;
+
+        if e_idx == -1 {
+            digits.push_str(&valid_float[dot_idx as usize + 1..]);
+        } else {
+            digits.push_str(&valid_float[dot_idx as usize + 1..e_idx as usize]);
+        }
+    }
+    if e_idx != -1 {
+        let exp = box_try!((&valid_float[e_idx as usize + 1..]).parse::<i64>());
+        if exp > 0 && int_cnt > (i64::MAX - exp) {
+            // (exp + incCnt) overflows MaxInt64.
+            return Err(box_err!("[1264] Data Out of Range"));
+        }
+        int_cnt += exp;
+    }
+    if int_cnt <= 0 || (int_cnt == 1 && (digits.starts_with('-') || digits.starts_with('+'))) {
+        return Ok("0".to_owned());
+    }
+    let mut valid_int = String::new();
+    if int_cnt <= digits.len() as i64 {
+        valid_int.push_str(&digits[..int_cnt as usize]);
+    } else {
+        let extra_zero_count = int_cnt - digits.len() as i64;
+        if extra_zero_count > 20 {
+            // Return overflow to avoid allocating too much memory.
+            return Err(box_err!("[1264] Data Out of Range"));
```

Comment:
Mysql error code.

This `box_err!("[1264] Data Out of Range")` is a temporary solution, I will refactor it later.